GATKPathSpecifier URI class and update FeatureInput.

cmnbroad opened the pull request:
URI class, fixes https://github.com/broadinstitute/gatk/issues/4343 for feature inputs.

cmnbroad commented:
@bbimber FYI - when this PR goes in I think it will require some minor changes to VariantQC if you use tagging syntax.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureInput.java:
```diff
@@ -321,7 +192,11 @@ public boolean equals(final Object other) {
 
         final FeatureInput<?> otherFeature = (FeatureInput<?>)other;
-        return name.equals(otherFeature.name) && featureFile.equals(otherFeature.featureFile);
+        if (!Objects.equals(getTag(), otherFeature.getTag())) {
```
Yeah, good catch.

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/engine/PathSpecifierUnitTest.java:
```diff
@@ -0,0 +1,353 @@
+                {"file:///path/to/localFile.bam",   "file:///path/to/localFile.bam",    true, true},  // empty authority
+
+                //*****************************************************************************
+                // Valid URIs which are NOT valid NIO paths (no installed file system provider)
```
Sigh. True, but its because only I have multiple branches with this code in different repos, and I'm trying to keep them as in-sync as possible in order to make propagating code review changes easier. In htjsdk-next, there are no providers for these so they're not valid there; when I moved them here I just changed the expected boolean to reflect that to make them pass. I'll have to reconcile them.

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/engine/PathSpecifierUnitTest.java:
```diff
@@ -0,0 +1,353 @@
+
+public class PathSpecifierUnitTest {
+
+    final static String FS_SEPARATOR = FileSystems.getDefault().getSeparator();
```
Yes.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/GATKPathSpecifier.java:
```diff
@@ -0,0 +1,162 @@
+                Path p = BucketUtils.getPathOnGcs(getURIString());
+                inputStream = Files.newInputStream(p);
+            } else if (getURI().getScheme().equals(HDFS_SCHEME)) {
+                org.apache.hadoop.fs.Path file = new org.apache.hadoop.fs.Path(getURIString());
```
Good question. I included this because its what `BucketUtils.openFile` does (and thus what I needed for my bigger version of this branch where I used the URI class for all reads and reference inputs). But maybe it isn't necessary ? Somehow I thought the jsr203hadoop provider was flake, but I'm not sure where I got that from.

cmnbroad replied:
Good question. I included this because its what `BucketUtils.openFile` does (and thus what I needed for my bigger version of this branch where I used the URI class for all reads and reference inputs). But maybe it isn't necessary ? Somehow I thought the jsr203hadoop provider was flaky, but I'm not sure where I got that from.

lbergelson replied:
It might be flakey, we haven't done any sort of stress testing on it and it's not really maintained as far as I know.  My guess is that's it's fine for getting a stream though.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathSpecifier.java:
```diff
@@ -0,0 +1,260 @@
+ *
+ * Definitions:
+ *
+ * "absolute" URI  - specifies a scheme
```
It is right from the URI RFC, and in my opinion thoroughly confusing. In the beginning I was constantly referring to it, which is why I included it here. Might be overkill though.

lbergelson replied:
I think it's worth including if it's the standard nomenclature, it's pretty unintuitive and confusing and having it here is useful in trying to understand things.  Maybe include a note about which RFC it's defined in.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/GATKPathSpecifier.java:
```diff
@@ -0,0 +1,162 @@
+
+        try {
+            InputStream inputStream;
+            if (getURI().getScheme().equals(GCS_SCHEME)) {
```
Yeah, I think you're right, for the GCS one.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/GATKPathSpecifier.java:
```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+    //TODO: should this wrap the stream in a .gz in a BlockCompressedInputStream like IOUtils does?
```
Agreed - this comment is pretty old.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/GATKPathSpecifier.java:
```diff
@@ -0,0 +1,162 @@
+
+        try {
+            OutputStream outputStream;
+            if (getURI().getScheme().equals(GCS_SCHEME)) {
```
Right.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/GATKPathSpecifier.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    @Override
+    public void setTag(String tagName) {
```
Yes. But its the CLP that calls them.

lbergelson replied:
I know... I'm just complaining without offering a better solution.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureInput.java:
```diff
@@ -30,35 +26,26 @@
      * File containing Features as specified by the user on the command line
      */
-    private final String featureFile;
+    private final String rawInputSpecifier;
```
Yes.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathSpecifier.java:
```diff
@@ -0,0 +1,260 @@
+
+    private final String    rawInputString;     // raw input string provided by th user; may or may not have a scheme
+    private final URI       uri;                // working URI; always has a scheme (assume "file" if not provided)
+    private String          pathFailureReason;  // cache the reason for "toPath" conversion failure
```
Makes sense.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/GATKPathSpecifier.java:
```diff
@@ -0,0 +1,162 @@
+
+        GATKPathSpecifier GATKPathSpecifier = (GATKPathSpecifier) o;
+
+        if (!Objects.equals(tagName, GATKPathSpecifier.tagName)) return false;
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathSpecifier.java:
```diff
@@ -0,0 +1,260 @@
+ *
+ * 1) PathSpecifier constructor: requires a syntactically valid URI, possibly containing a scheme (if no scheme
+ *    is present the path part will be escaped/encoded), or a valid local file reference
+ * 2) isNio: true if the input string is an identifier that is syntactically valid, and is backed by
```
What do you think about getting rid of this completely for now ? Most of the cases that I know of will branch based on the scheme, otherwise they assume its Path-able. I'm still not sure there is a use-case where something will call this and do something useful when it returns false.

lbergelson replied:
We could remove it... I'm also not totally clear on the use case.  Maybe for displaying an error message?

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathSpecifier.java:
```diff
@@ -0,0 +1,260 @@
+        if (getCachedPath() != null) {
+            return getCachedPath();
+        } else {
+            final Path tmpPath = Paths.get(getURI());
```
I agree, if we back port it. But for GATK, I don't think we want this to use IOUtils as that can create bogus Paths (mentioned in the comments I added to IOUtils). As we migrate code in GATK to use this class, keeping it separate will ensure we don't introduce new dependencies on that old behavior.

lbergelson replied:
@cmnbroad I believe there's a pretty serious issue with this here still.  There's an important spark bug that's solved by these lines in `IOUtils.getPath`


```
  } catch (FileSystemNotFoundException e) {
            try {
                ClassLoader cl = Thread.currentThread().getContextClassLoader();
                if ( cl == null ) {
                    throw e;
                }
                return FileSystems.newFileSystem(uri, new HashMap<>(), cl).provider().getPath(uri);
            } 
....
```

The default filesystem loading uses the system class path which is set by spark before loading the client jars.  These lines cause it to try again using the classloader that includes the client jars.  I think we should continue to do this in this implementation of getPath.  What we don't have to do is the fallback to treating non-interpretable paths as local Files.

cmnbroad replied:
@lbergelson I see - I'll have to incorporate that somehow in GATKPathSpecifier then.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathSpecifier.java:
```diff
@@ -0,0 +1,260 @@
+
+        PathSpecifier that = (PathSpecifier) o;
+
+        if (!getURIString().equals(that.getURIString())) return false;
```
I think I agree. I have mixed feelings since different input strings can result in the same URI. But I think you're right that the stricter equality is probably better.

lbergelson replied:
I meant more that it was repetitive to compare the uriString AND the URI itself since those should hopefully be equivalent?  It might make sense to have equality be on the normalized URI and ignore the raw String.  It might also make sense not to define equality if we can't decide what it means unambiguously.

cmnbroad replied:
Right - it is repetitive.`FeatureManager` for example uses `FeatureInputs` as `Map` keys,  so I think we should put some stake in the ground about equality. My original implementation would have treated different input strings that resolve to the same URI as the equal, but I think the stricter implementation (treat them as different) makes more sense. If some caller wants to use URI equality, it do so with the use the URIs ?

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathURI.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Interface for htsjdk input/output URIs.
+ */
+public interface PathURI {
```
I THINK its more of an implementation detail than something that affects the interface itself. I originally included it in the implementation class because I had to keep referring back to it when I first started coding that, and was confused about some of the terminology.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java:
```diff
@@ -778,6 +781,9 @@ public static Path getPath(String uriString) {
             catch (ProviderNotFoundException x) {
+                // TODO: this creates bogus Path on the current file system for schemes such as gendb, nonexistent, gcs
+                // TODO: we depend on this code path to allow IntervalUtils to all getPath on a string that may be either
+                // a literal interval or a feature file containing intervals
```
Yes. Hopefully the next PR will include fixing those code paths to not rely on this behavior, and then eventually we can get rid of this altogether.

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/engine/PathSpecifierUnitTest.java:
```diff
@@ -0,0 +1,353 @@
+
+                //*****************************************************************************************
+                // Reference that contain characters that require URI-encoding. If the input string is presented
+                // without no scheme, it will be be automatically encoded by PathSpecifier, otherwise it
```
Done.

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/engine/PathSpecifierUnitTest.java:
```diff
@@ -0,0 +1,353 @@
+    @Test
+    public void testStdIn() throws IOException {
+        final PathURI htsURI = new PathSpecifier(
+                SystemUtils.IS_OS_WINDOWS ?
```
This particular code is actually bogus; some Windows apps use "-" to mean "stdout", but its not standard or built-in to the file system namespace, so this doesn't work in most cases. I've changed this to throw a `SkipException` on Windows.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/PathSpecifier.java:
```diff
@@ -0,0 +1,260 @@
+
+    @Override
+    public int hashCode() {
+        int result = getURIString().hashCode();
```
I THINK these are in right now (or at least hashCode and equals agree).

cmnbroad commented:
I think I responded to everything. Renamed isNIO to hasFileSystemProvider and keeping it for now. back to @lbergelson.

cmnbroad commented:
@lbergelson I added the ClassLoader fallback code in to `GATKPathSpecifier`. In doing so, I expected some of the `PathSpecifier` negative unit tests to produce different exceptions, but they didn't. This was because the tests were all using the `PathSpecifier` base class (because I took them from htsjdk-next which has only the base implementation). So I updated the tests to use `GATKPathSpecifier` (they all passed), then integrated the fallback code and changed 3 negative tests that went from FileSystemNotFoundException to ProviderNotFoundException. Finally I renamed the test class file to correctly reflect the class being tested. Sorry this was such a pain.

cmnbroad commented:
Also, once the tests pass, I'm going to rebase on master and then run them one more time.
